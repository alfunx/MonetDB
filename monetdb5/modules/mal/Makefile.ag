# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0.  If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# Copyright 1997 - July 2008 CWI, August 2008 - 2016 MonetDB B.V.

INCLUDES = ../../mal ../atoms ../kernel \
	../../../clients/mapilib \
	../../../common/options \
	../../../common/stream \
	../../../common/utils \
	../../../gdk \
	$(pcre_CFLAGS) $(zlib_CFLAGS) $(BZIP_INCS) $(MSGCONTROL_FLAGS) \
	$(openssl_CFLAGS)

MTSAFE

lib_mal = {
	NOINST
	SOURCES = \
		batExtensions.c batExtensions.h \
		bbp.c bbp.h \
		clients.c clients.h \
		factories.c factories.h \
		groupby.c groupby.h \
		inspect.c inspect.h \
		iterator.c  iterator.h \
		projectionpath.c  projectionpath.h \
		language.c language.h \
		mal_io.c mal_io.h \
		mal_mapi.c mal_mapi.h \
		manual.c manual.h \
		mat.c mat.h \
		mdb.c mdb.h \
		mkey.c mkey.h \
		manifold.c manifold.h \
		pcre.c \
		profiler.c profiler.h \
		querylog.c querylog.h \
		recycle.c recycle.h \
		remote.c remote.h \
		sabaoth.c sabaoth.h \
		sysmon.c sysmon.h \
		tablet.c tablet.h \
		tokenizer.c tokenizer.h \
		transaction.c \
		txtsim.c txtsim.h \
		zorder.c zorder.h \
		sample.c sample.h \
		json_util.c json_util.h \
		calc.c batcalc.c
}

headers_mal = {
	HEADERS = mal
	DIR = libdir/monetdb5
	SOURCES = language.mal mal_init.mal bbp.mal \
		profiler.mal batExtensions.mal \
		inspect.mal manual.mal mal_io.mal mkey.mal manifold.mal \
		iterator.mal clients.mal \
		factories.mal groupby.mal mdb.mal pcre.mal mat.mal \
		transaction.mal \
		mal_mapi.mal sabaoth.mal remote.mal  \
		txtsim.mal recycle.mal \
		tokenizer.mal zorder.mal sample.mal json_util.mal \
		batmtime.mal querylog.mal sysmon.mal
}

headers_auto = {
	HEADERS = mal
	DIR = libdir/monetdb5/autoload
	SOURCES = 01_calc.mal 01_batcalc.mal
}

headers_hge = {
	COND = HAVE_HGE
	HEADERS = mal
	DIR = libdir/monetdb5/autoload
	SOURCES = 00_calc_hge.mal 00_batcalc_hge.mal \
		00_iterator_hge.mal 00_language_hge.mal \
		00_mal_mapi_hge.mal 00_mkey_hge.mal 00_batExtensions_hge.mal
}

EXTRA_DIST = batExtensions.mal iterator.mal \
	groupby.mal mal_init.mal manual.mal mkey.mal manifold.mal pcre.mal \
	profiler.mal recycle.mal remote.mal sabaoth.mal \
	transaction.mal txtsim.mal tablet.mal tablet.h sample.mal json_util.mal \
	mal_mapi.mal mat.mal tokenizer.mal \
	batmtime.mal querylog.mal sysmon.mal \
	01_calc.mal 01_batcalc.mal \
	00_calc_hge.mal 00_batcalc_hge.mal \
	00_iterator_hge.mal 00_language_hge.mal \
	00_mal_mapi_hge.mal 00_mkey_hge.mal 00_batExtensions_hge.mal

EXTRA_DIST_DIR = Tests
