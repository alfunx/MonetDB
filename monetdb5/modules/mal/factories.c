/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright 1997 - July 2008 CWI, August 2008 - 2016 MonetDB B.V.
 */

/*
 * (c) Martin Kersten
 * The factory infrastructure can be inspected and steered with
 * the commands provided here. to-be-completed-when-needed
 */
#include "monetdb_config.h"
#include "factories.h"

str
FCTgetPlants(bat *ret, bat *ret2)
{
	(void) ret;
	(void) ret2;
	throw(MAL, "factories.getPlants", PROGRAM_NYI);
}

str
FCTgetCaller(int *ret)
{
	(void) ret;
	throw(MAL, "factories.getCaller", PROGRAM_NYI);
}

str
FCTgetOwners(bat *ret)
{
	(void) ret;
	throw(MAL, "factories.getOwner", PROGRAM_NYI);
}

str
FCTgetArrival(bat *ret)
{
	(void) ret;
	throw(MAL, "factories.getArrival", PROGRAM_NYI);
}

str
FCTgetDeparture(bat *ret)
{
	(void) ret;
	throw(MAL, "factories.getDeparture", PROGRAM_NYI);
}

str
FCTshutdown(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci)
{
	str mod = *getArgReference_str(stk, pci, 1);
	str fcn = *getArgReference_str(stk, pci, 2);
	Symbol s;
	(void) mb;

	s = findSymbol(cntxt->nspace, putName(mod), putName(fcn));
	if (s == NULL)
		throw(MAL, "factories.shutdown", RUNTIME_OBJECT_MISSING);
	shutdownFactory(cntxt,s->def);
	return MAL_SUCCEED;
}
