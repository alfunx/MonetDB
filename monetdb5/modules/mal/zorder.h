/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright 1997 - July 2008 CWI, August 2008 - 2016 MonetDB B.V.
 */

#ifndef _ZORDER_H
#define _ZORDER_H


mal_export str ZORDencode_int_oid(oid *z, int *x, int *y);
mal_export str ZORDbatencode_int_oid(bat *z, bat *x, bat *y);
mal_export str ZORDdecode_int_oid(int *x, int *y, oid *z);
mal_export str ZORDdecode_int_oid_x(int *x, oid *z);
mal_export str ZORDdecode_int_oid_y(int *y, oid *z);
mal_export str ZORDbatdecode_int_oid(bat *x, bat *y, bat *z);
mal_export str ZORDbatdecode_int_oid_x(bat *x, bat *z);
mal_export str ZORDbatdecode_int_oid_y(bat *y, bat *z);
mal_export str ZORDslice_int(bat *r, int *xb, int *yb, int *xt, int *yt);

#endif /* _ZORDER_H */
