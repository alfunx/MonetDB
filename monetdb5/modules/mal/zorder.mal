# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0.  If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# Copyright 1997 - July 2008 CWI, August 2008 - 2016 MonetDB B.V.

module zorder;

command encode(x:int,y:int):oid
address ZORDencode_int_oid
comment "Derive the z-order index from a value pair";

command encode(x:bat[:int],y:bat[:int]):bat[:oid]
address ZORDbatencode_int_oid
comment "Derive the z-order index from a value pair";

command decode(z:oid)(x:int,y:int)
address ZORDdecode_int_oid
comment "Derive the z-order pair";

command decode(z:bat[:oid])(:bat[:int],:bat[:int])
address ZORDbatdecode_int_oid
comment "Derive the z-order pair";

command decode_x(z:oid)(x:int)
address ZORDdecode_int_oid_x
comment "Derive the z-order x-coordinate";

command decode_x(z:bat[:oid]):bat[:int]
address ZORDbatdecode_int_oid_x
comment "Derive the z-order x-coordinate";

command decode_y(z:oid)(y:int)
address ZORDdecode_int_oid_y
comment "Derive the z-order y-coordinate";

command decode_y(z:bat[:oid]):bat[:int]
address ZORDbatdecode_int_oid_y
comment "Derive the z-order y-coordinate";

command slice(xb:int, yb:int, xt:int, yt:int ):bat[:oid]
address ZORDslice_int
comment "Extract the Z-order indices between two points";
