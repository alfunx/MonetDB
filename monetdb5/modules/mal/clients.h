/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright 1997 - July 2008 CWI, August 2008 - 2016 MonetDB B.V.
 */

/*
 * @+ Monet client data
 * Portions of the client record can be directly obtained for
 * backward compatibility. The routine clientInfo provides more
 * detailed information.
 */
#ifndef _CLIENTS_H
#define _CLIENTS_H
#include "mal.h"
#include "mal_interpreter.h"

mal_export str CLTsetListing(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTgetClientId(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTgetScenario(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTsetScenario(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTusers(bat *ret);
mal_export str CLTsetHistory(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTquit(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTLogin(bat *ret, bat *nme);
mal_export str CLTLastCommand(bat *ret);
mal_export str CLTActions(bat *ret);
mal_export str CLTTime(bat *ret);
mal_export str CLTInfo(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTstop(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTsuspend(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTsetTimeout(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTsetSessionTimeout(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTgetTimeout(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTwakeup(void *ret, int *id);

mal_export str CLTshutdown(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTmd5sum(str *ret, str *pw);
mal_export str CLTsha1sum(str *ret, str *pw);
mal_export str CLTripemd160sum(str *ret, str *pw);
mal_export str CLTsha2sum(str *ret, str *pw, int *bits);
mal_export str CLTbackendsum(str *ret, str *pw);
mal_export str CLTaddUser(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTremoveUser(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTgetUsername(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTgetPasswordHash(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTchangeUsername(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTchangePassword(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTsetPassword(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTcheckPermission(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTgetUsers(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str CLTsessions(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
#endif /* _CLIENTS_H */
