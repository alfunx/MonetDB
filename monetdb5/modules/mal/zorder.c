/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright 1997 - July 2008 CWI, August 2008 - 2016 MonetDB B.V.
 */

/*
 * author Martin Kersten
 * Z-order
 * This module provides the primitives to implement 2-dim Z-order functionality.
 * Arrays stored in Z-order have a better locality of reference for many
 * operations. Slicing part of the array amounts to deriving a BAT with
 * the z-order indices required, whereafter a simple semijoin would be
 * sufficient to fetch the values.
 *
 * The encoding produces OIDs, which makes it easy to align
 * any void headed BAT as a sorted Z-ordered representation.
 * This gives both fast point access and clustered slicing.
 */

#include "monetdb_config.h"
#include "mal.h"
#include "mal_exception.h"
#include "zorder.h"

static inline oid Zencode_int_oid(int x, int y)
{
	oid v = 0;
	int i,mask=1;
	for ( i = 0; i < (int) (8 * sizeof(oid)/2) ; i++) {
		v |= ((x & 1) * mask);
		x>>=1;
		mask <<= 1;
		v |= ((y & 1) * mask);
		y>>=1;
		mask <<= 1;
	}
	return v;
}

static inline void Zdecode_int_oid(int *x, int *y, oid *z)
{
	int xv = 0, yv=0, mask =1;
	oid zv = *z;
	int i;
	for ( i = 0; i < (int) (8 * sizeof(oid)); i+= 2) {
		xv |= ((zv & 1) * mask);
		zv >>= 1;
		yv |= ((zv & 1) * mask);
		zv >>= 1;
		mask <<=1;
	}
	*x = xv;
	*y = yv;
}
static inline void Zdecode_int_oid_x(int *x, oid *z)
{
	int xv = 0, mask =1;
	oid zv = *z;
	int i;
	for ( i = 0; i < (int) (8 * sizeof(oid)); i+= 2) {
		xv |= ((zv & 1) * mask);
		zv >>= 2;
		mask <<=1;
	}
	*x = xv;
}
static inline void Zdecode_int_oid_y(int *y, oid *z)
{
	int yv=0, mask =1;
	oid zv = *z;
	int i;
	for ( i = 0; i < (int) (8 * sizeof(oid)); i+= 2) {
		zv >>= 1;
		yv |= ((zv & 1) * mask);
		zv >>= 1;
		mask <<=1;
	}
	*y = yv;
}

str 
ZORDencode_int_oid(oid *z, int *x, int *y)
{
	*z = Zencode_int_oid(*x,*y);
	return MAL_SUCCEED;
}

str 
ZORDdecode_int_oid(int *x, int *y, oid *z)
{
	Zdecode_int_oid(x,y,z);
	return MAL_SUCCEED;
}

str 
ZORDdecode_int_oid_x(int *x, oid *z)
{
	Zdecode_int_oid_x(x,z);
	return MAL_SUCCEED;
}

str 
ZORDdecode_int_oid_y(int *y, oid *z)
{
	Zdecode_int_oid_y(y,z);
	return MAL_SUCCEED;
}

str
ZORDbatencode_int_oid(bat *zbid, bat *xbid, bat *ybid)
{
	BAT *bx, *by,*bz;
	int *p, *q, *r;
	oid *z;

	bx = BATdescriptor(*xbid);
	by = BATdescriptor(*ybid);
	if ( bx == 0 || by == 0){
		if ( bx ) BBPunfix(bx->batCacheid);
		if ( by ) BBPunfix(by->batCacheid);
		throw(OPTIMIZER, "zorder.encode", RUNTIME_OBJECT_MISSING);
	}
	if ( BATcount(bx) != BATcount(by)){
		BBPunfix(bx->batCacheid);
		BBPunfix(by->batCacheid);
		throw(OPTIMIZER, "zorder.encode", ILLEGAL_ARGUMENT);
	}
	
	bz = BATnew(TYPE_void, TYPE_oid, BATcount(bx), TRANSIENT);
	if (bz == 0){
		BBPunfix(bx->batCacheid);
		BBPunfix(by->batCacheid);
		throw(OPTIMIZER, "zorder.encode", MAL_MALLOC_FAIL);
	}
	p = (int *) Tloc(bx, BUNfirst(bx));
	q = (int *) Tloc(bx, BUNlast(bx));
	r = (int *) Tloc(by, BUNfirst(by));
	z = (oid *) Tloc(bz, BUNfirst(bz));

	if ( bx->T->nonil && by->T->nonil){
		for ( ; p<q; z++,p++,r++)
			*z = Zencode_int_oid( *p, *r );
	} else
	if ( bx->T->nonil ){
		for ( ; p<q; z++,p++,r++)
		if ( *r == int_nil)
			*z = oid_nil;
		else
			*z = Zencode_int_oid( *p, *r );
	} else
	if ( by->T->nonil ){
		for ( ; p<q; z++,p++,r++)
		if ( *p == int_nil)
			*z = oid_nil;
		else
			*z = Zencode_int_oid( *p, *r );
	} else {
		for ( ; p<q; z++,p++,r++)
		if ( *r == int_nil)
			*z = oid_nil;
		else
		if ( *p == int_nil)
			*z = oid_nil;
		else
			*z = Zencode_int_oid( *p, *r );
	}

	BBPunfix(bx->batCacheid);
	BBPunfix(by->batCacheid);

	if (!(bz->batDirty&2)) 
		BATsetaccess(bz, BAT_READ);
	BATsetcount(bz, BATcount(bx));
	BATseqbase(bz, bx->hseqbase);
	bz->hsorted = 1;
	bz->hrevsorted = 0;
	bz->tsorted = 0;
	bz->trevsorted = 0;
	bz->H->nonil = 1;
	bz->T->nonil = bx->T->nonil && by->T->nonil;

	BBPkeepref(*zbid = bz->batCacheid);
	return MAL_SUCCEED;
}

str
ZORDbatdecode_int_oid(bat *xbid, bat *ybid, bat *zbid)
{
	BAT *bx, *by,*bz;
	oid *z, *q;
	int *x, *y;

	bz = BATdescriptor(*zbid);
	if ( bz == 0 )
		throw(OPTIMIZER, "zorder.decode", RUNTIME_OBJECT_MISSING);
	
	bx = BATnew(TYPE_void, TYPE_int, BATcount(bz), TRANSIENT);
	by = BATnew(TYPE_void, TYPE_int, BATcount(bz), TRANSIENT);
	if ( bx == 0 || by == 0 ){
		if ( bx ) BBPunfix(bx->batCacheid);
		if ( by ) BBPunfix(by->batCacheid);
		BBPunfix(bz->batCacheid);
		throw(OPTIMIZER, "zorder.decode", RUNTIME_OBJECT_MISSING);
	}
	
	z = (oid *) Tloc(bz, BUNfirst(bz));
	q = (oid *) Tloc(bz, BUNlast(bz));
	x = (int *) Tloc(bx, BUNfirst(bx));
	y = (int *) Tloc(by, BUNfirst(by));

	if ( bz->T->nonil ){
		for ( ; z<q; z++,x++,y++)
			Zdecode_int_oid(x,y,z);
	} else {
		for ( ; z<q; z++,x++,y++)
		if ( *z == oid_nil) {
			*x = int_nil;
			*y = int_nil;
		} else
			Zdecode_int_oid( x,y,z);
	} 

	if (!(bx->batDirty&2)) 
		BATsetaccess(bx, BAT_READ);
	BATsetcount(bx, BATcount(bz));
	BATseqbase(bx, bz->hseqbase);
	bx->hsorted = 1;
	bx->hrevsorted = 0;
	bx->tsorted = 0;
	bx->trevsorted = 0;
	bx->H->nonil = 1;
	bx->T->nonil = bz->T->nonil;

	if (!(by->batDirty&2)) 
		BATsetaccess(by, BAT_READ);
	BATsetcount(by, BATcount(bz));
	BATseqbase(by, bz->hseqbase);
	by->hsorted = 1;
	by->hrevsorted = 0;
	by->tsorted = 0;
	by->trevsorted = 0;
	by->H->nonil = 1;
	by->T->nonil = bz->T->nonil;

	BBPunfix(bz->batCacheid);
	BBPkeepref(*xbid = bx->batCacheid);
	BBPkeepref(*ybid = by->batCacheid);
	return MAL_SUCCEED;
}

str
ZORDbatdecode_int_oid_x(bat *xbid, bat *zbid)
{
	BAT *bx,*bz;
	oid *z, *q;
	int *x;

	bz = BATdescriptor(*zbid);
	if ( bz == 0 )
		throw(OPTIMIZER, "zorder.decode", RUNTIME_OBJECT_MISSING);
	
	bx = BATnew(TYPE_void, TYPE_int, BATcount(bz), TRANSIENT);
	if ( bx == 0 ){
		BBPunfix(bz->batCacheid);
		throw(OPTIMIZER, "zorder.decode", RUNTIME_OBJECT_MISSING);
	}
	
	z = (oid *) Tloc(bz, BUNfirst(bz));
	q = (oid *) Tloc(bz, BUNlast(bz));
	x = (int *) Tloc(bx, BUNfirst(bx));

	if ( bz->T->nonil ){
		for ( ; z<q; z++,x++)
			Zdecode_int_oid_x(x,z);
	} else {
		for ( ; z<q; z++,x++)
		if ( *z == oid_nil) {
			*x = int_nil;
		} else
			Zdecode_int_oid_x(x,z);
	} 

	if (!(bx->batDirty&2)) 
		BATsetaccess(bx, BAT_READ);
	BATsetcount(bx, BATcount(bz));
	BATseqbase(bx, bz->hseqbase);
	bx->hsorted = 1;
	bx->hrevsorted = 0;
	bx->tsorted = 0;
	bx->trevsorted = 0;
	bx->H->nonil = 1;
	bx->T->nonil = bz->T->nonil;

	BBPunfix(bz->batCacheid);
	BBPkeepref(*xbid = bx->batCacheid);
	return MAL_SUCCEED;
}

str
ZORDbatdecode_int_oid_y(bat *ybid, bat *zbid)
{
	BAT *by,*bz;
	oid *z, *q;
	int *y;

	bz = BATdescriptor(*zbid);
	if ( bz == 0 )
		throw(OPTIMIZER, "zorder.decode", RUNTIME_OBJECT_MISSING);
	
	by = BATnew(TYPE_void, TYPE_int, BATcount(bz), TRANSIENT);
	if ( by == 0 ){
		BBPunfix(bz->batCacheid);
		throw(OPTIMIZER, "zorder.decode", RUNTIME_OBJECT_MISSING);
	}
	
	z = (oid *) Tloc(bz, BUNfirst(bz));
	q = (oid *) Tloc(bz, BUNlast(bz));
	y = (int *) Tloc(by, BUNfirst(by));

	if ( bz->T->nonil ){
		for ( ; z<q; z++,y++)
			Zdecode_int_oid_y(y,z);
	} else {
		for ( ; z<q; z++,y++)
		if ( *z == oid_nil) {
			*y = int_nil;
		} else
			Zdecode_int_oid_y(y,z);
	} 

	if (!(by->batDirty&2)) 
		BATsetaccess(by, BAT_READ);
	BATsetcount(by, BATcount(bz));
	BATseqbase(by, bz->hseqbase);
	by->hsorted = 1;
	by->hrevsorted = 0;
	by->tsorted = 0;
	by->trevsorted = 0;
	by->H->nonil = 1;
	by->T->nonil = bz->T->nonil;

	BBPunfix(bz->batCacheid);
	BBPkeepref(*ybid = by->batCacheid);
	return MAL_SUCCEED;
}

str ZORDslice_int(bat *r, int *xb, int *yb, int *xt, int *yt)
{
	BAT *bn;
	int i,j;
	oid zv;

	bn = BATnew(TYPE_void, TYPE_oid, 0, TRANSIENT);
	if( bn == 0)
		throw(OPTIMIZER, "zorder.slice", MAL_MALLOC_FAIL);
	BATseqbase(bn, 0);
	/* use the expensive road, could be improved by bit masking */
	for ( i= *xb; i < *xt; i++)
	{
		for (j= *yb; j < *yt; j++){
			zv= Zencode_int_oid(i,j);
			BUNappend(bn, &zv, FALSE);
		}
	}

	if (!(bn->batDirty&2)) 
		BATsetaccess(bn, BAT_READ);
	BBPkeepref(*r = bn->batCacheid);
	return MAL_SUCCEED;
}
