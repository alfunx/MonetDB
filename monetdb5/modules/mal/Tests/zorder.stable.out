stdout of test 'zorder` in directory 'monetdb5/modules/mal` itself:


# 08:41:27 >  
# 08:41:27 >   mserver5 "--config=/ufs/mk/current//Linux/etc/monetdb5.conf" --debug=10 --set gdk_nr_threads=0 --set "monet_mod_path=/ufs/mk/current//Linux/lib/MonetDB5:/ufs/mk/current//Linux/lib/MonetDB5/lib:/ufs/mk/current//Linux/lib/MonetDB5/bin" --set "gdk_dbfarm=/ufs/mk/current//Linux/var/MonetDB5/dbfarm"  --set mapi_open=true --set xrpc_open=true --set mapi_port=33081 --set xrpc_port=46016 --set monet_prompt= --set mal_listing=2 --trace  --dbname=mTests_src_modules_mal  zorder.mal
# 08:41:27 >  

# MonetDB server v5.19.0, based on kernel v1.37.0
# Serving database 'mTests_src_modules_mal', using 4 threads
# Compiled for x86_64-unknown-linux-gnu/64bit with 64bit OIDs dynamically linked
# Found 7.751 GiB available main-memory.
# Copyright (c) 1993-July 2008 CWI.
# Copyright (c) August 2008-2015 MonetDB B.V., all rights reserved
# Visit http://www.monetdb.org/ for further information
# Listening for connection requests on mapi:monetdb://vienna.da.cwi.nl:35410/
# Listening for UNIX domain connection requests on mapi:monetdb:///var/tmp/mtest-26774/.s.monetdb.35410
# MonetDB/GIS module loaded
# Start processing logs sql/sql_logs version 52200
# Start reading the write-ahead log 'sql_logs/sql/log.4'
# Finished reading the write-ahead log 'sql_logs/sql/log.4'
# Finished processing logs sql/sql_logs
# MonetDB/SQL module loaded
# MonetDB/R   module loaded

Ready.

# 09:46:33 >  
# 09:46:33 >  "mclient" "-lmal" "-ftest" "-Eutf-8" "--host=/var/tmp/mtest-26774" "--port=35410"
# 09:46:33 >  

[ 0@0 ]
[ 1@0 ]
[ 2@0 ]
[ 3@0 ]
[ 4@0 ]
[ 5@0 ]
[ 6@0 ]
[ 7@0 ]
[ 8@0 ]
[ 9@0 ]
[ 10@0 ]
[ 11@0 ]
[ 0,	2	]
[ 1,	2	]
[ 0,	3	]
[ 1,	3	]
[ 0,	2	]
[ 1,	2	]
[ 0,	3	]
[ 1,	3	]
#--------------------------#
# t	t	t  # name
# void	int	int	  # type
#-------------------------#
[ 0@0,	  0,	  0	  ]
[ 1@0,	  1,	  0	  ]
[ 2@0,	  0,	  1	  ]
[ 3@0,	  1,	  1	  ]
[ 4@0,	  2,	  0	  ]
[ 5@0,	  3,	  0	  ]
[ 6@0,	  2,	  1	  ]
[ 7@0,	  3,	  1	  ]
[ 8@0,	  0,	  2	  ]
[ 9@0,	  1,	  2	  ]
[ 10@0,	  0,	  3	  ]
[ 11@0,	  1,	  3	  ]
[ 12@0,	  2,	  2	  ]
[ 13@0,	  3,	  2	  ]
[ 14@0,	  2,	  3	  ]
[ 15@0,	  3,	  3	  ]
#-----------------#
# h	t	  # name
# void	oid	  # type
#-----------------#
[ 0@0,	  0@0	  ]
[ 1@0,	  1@0	  ]
[ 2@0,	  2@0	  ]
[ 3@0,	  3@0	  ]
[ 4@0,	  4@0	  ]
[ 5@0,	  5@0	  ]
[ 6@0,	  6@0	  ]
[ 7@0,	  7@0	  ]
[ 8@0,	  8@0	  ]
[ 9@0,	  9@0	  ]
[ 10@0,	  10@0	  ]
[ 11@0,	  11@0	  ]
[ 12@0,	  12@0	  ]
[ 13@0,	  13@0	  ]
[ 14@0,	  14@0	  ]
[ 15@0,	  15@0	  ]
#-------------------------#
# h	t	t	  # name
# void	int	int	  # type
#-------------------------#
[ 0@0,	  0,	  0	  ]
[ 1@0,	  1,	  0	  ]
[ 2@0,	  0,	  1	  ]
[ 3@0,	  1,	  1	  ]
[ 4@0,	  2,	  0	  ]
[ 5@0,	  3,	  0	  ]
[ 6@0,	  2,	  1	  ]
[ 7@0,	  3,	  1	  ]
[ 8@0,	  0,	  2	  ]
[ 9@0,	  1,	  2	  ]
[ 10@0,	  0,	  3	  ]
[ 11@0,	  1,	  3	  ]
[ 12@0,	  2,	  2	  ]
[ 13@0,	  3,	  2	  ]
[ 14@0,	  2,	  3	  ]
[ 15@0,	  3,	  3	  ]
#-------------------------#
# h	t	t	  # name
# void	int	int	  # type
#-------------------------#
[ 0@0,	  0,	  0	  ]
[ 1@0,	  1,	  0	  ]
[ 2@0,	  0,	  1	  ]
[ 3@0,	  1,	  1	  ]
[ 4@0,	  2,	  0	  ]
[ 5@0,	  3,	  0	  ]
[ 6@0,	  2,	  1	  ]
[ 7@0,	  3,	  1	  ]
[ 8@0,	  0,	  2	  ]
[ 9@0,	  1,	  2	  ]
[ 10@0,	  0,	  3	  ]
[ 11@0,	  1,	  3	  ]
[ 12@0,	  2,	  2	  ]
[ 13@0,	  3,	  2	  ]
[ 14@0,	  2,	  3	  ]
[ 15@0,	  3,	  3	  ]
#-----------------#
# h	t	  # name
# void	oid	  # type
#-----------------#
[ 0@0,	  3@0	  ]
[ 1@0,	  9@0	  ]
[ 2@0,	  6@0	  ]
[ 3@0,	  12@0	  ]
#-------------------------#
# h	t	t	  # name
# void	int	int	  # type
#-------------------------#
[ 0@0,	  1,	  1	  ]
[ 1@0,	  1,	  2	  ]
[ 2@0,	  2,	  1	  ]
[ 3@0,	  2,	  2	  ]

# 08:41:28 >  
# 08:41:28 >  Done.
# 08:41:28 >  

